contracts/modules/capital/CowSwapOperator.sol

```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.9;
+
+import "../../external/cow/GPv2Order.sol";
+import "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts-v4/utils/math/Math.sol";
+import "../../interfaces/INXMMaster.sol";
+import "../../interfaces/IPool.sol";
+import "../../interfaces/IWeth.sol";
+import "../../interfaces/ICowSettlement.sol";
+import "../../interfaces/IPriceFeedOracle.sol";
+
+contract CowSwapOperator {
+  // Storage
+  ICowSettlement public immutable cowSettlement;
+  address public immutable cowVaultRelayer;
+  INXMMaster public immutable master;
+  address public immutable swapController;
+  IWeth public immutable weth;
+  IPriceFeedOracle public immutable priceFeedOracle;
+  bytes public currentOrderUID;
+  bytes32 public immutable domainSeparator;
+
+  // Constants
+  address public constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+  uint16 constant MAX_SLIPPAGE_DENOMINATOR = 10000;
+  uint256 constant MIN_VALID_TO_PERIOD = 600; // 10 minutes
+  uint256 constant MAX_VALID_TO_PERIOD = 3600; // 60 minutes
+  uint256 constant MIN_SELL_AMT_TO_FEE_RATIO = 100; // Sell amount at least 100x fee amount
+
+  event OrderPlaced(GPv2Order.Data order);
+  event OrderClosed(GPv2Order.Data order, uint256 filledAmount);
+
+  modifier onlyController() {
+    require(msg.sender == swapController, "SwapOp: only controller can execute");
+    _;
+  }
+
+  constructor(
+    address _cowSettlement,
+    address _swapController,
+    address _master,
+    address _weth,
+    address _priceFeedOracle
+  ) {
+    cowSettlement = ICowSettlement(_cowSettlement);
+    cowVaultRelayer = cowSettlement.vaultRelayer();
+    master = INXMMaster(_master);
+    swapController = _swapController;
+    weth = IWeth(_weth);
+    priceFeedOracle = IPriceFeedOracle(_priceFeedOracle);
+    domainSeparator = cowSettlement.domainSeparator();
+  }
+
+  receive() external payable {}
+
+  function getDigest(GPv2Order.Data calldata order) public view returns (bytes32) {
+    bytes32 hash = GPv2Order.hash(order, domainSeparator);
+    return hash;
+  }
+
+  function getUID(GPv2Order.Data calldata order) public view returns (bytes memory) {
+    bytes memory uid = new bytes(56);
+    bytes32 digest = getDigest(order);
+    GPv2Order.packOrderUidParams(uid, digest, order.receiver, order.validTo);
+    return uid;
+  }
+
+  function placeOrder(GPv2Order.Data calldata order, bytes calldata orderUID) public onlyController {
+    // Validate there's no current order going on
+    require(currentOrderUID.length == 0, "SwapOp: an order is already in place");
+
+    // Order UID verification
+    validateUID(order, orderUID);
+
+    // Validate basic CoW params
+    validateBasicCowParams(order);
+
+    // Validate feeAmount is not too high
+    require(order.sellAmount / order.feeAmount >= MIN_SELL_AMT_TO_FEE_RATIO, "SwapOp: Fee is above 1% of sellAmount");
```

Comment:
In the past I offered both options (percentage and absolute) to Hugh and he thought the 1% was fine. But nothing stops us from also adding an absolute amount. Althought we would have to use the oracle and increase the gas cost a bit. Up to you. About the revert message, you're right I'll change it.

Reply:
I'd suggest using errors instead of strings for revert reasons as they are cheaper and allow the message to be customized with the actual values:

https://docs.soliditylang.org/en/v0.8.9/structure-of-a-contract.html#errors